libcontainer/utils/utils_unix.go

```diff
@@ -67,3 +68,15 @@ func NewSockPair(name string) (parent *os.File, child *os.File, err error) {
 	}
 	return os.NewFile(uintptr(fds[1]), name+"-p"), os.NewFile(uintptr(fds[0]), name+"-c"), nil
 }
+
+// RunWithUmask runs a function f with umask mask.
+// It does so while locking the OS thread, as umask
+// operates on threads, and this avoid potentially leaking umask(0000)
+// to other threads.
+func RunWithUmask(mask int, f func() error) error {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+	oldMask := unix.Umask(mask)
+	defer unix.Umask(oldMask)
+	return f()
```

Comment:
The umask is part of the "filesystem information" whose sharing is governed by the `CLONE_FS` flag. [The Go runtime clones new threads with the `CLONE_FS` flag set](https://github.com/golang/go/blob/895664482c0ebe5cec4a6935615a1e9610bbf1e3/src/runtime/os_linux.go#L153) so any call to `unix.Umask()` immediately affects all goroutines in the process. `runtime.LockOSThread()` alone is not enough to prevent the change from leaking to other threads.

Here's what would work:
```suggestion
	errC := make(chan error)
	go func() {
		runtime.LockOSThread()
		if err := unix.Unshare(unix.CLONE_FS); err != nil {
			errC <- err
			runtime.UnlockOSThread()
			return
		}
		unix.Umask(mask)
		errC <- f()
	}()
	return <-errC
```

`unix.Unshare(CLONE_FS)` is an irreversible operation; the calling thread will never again have its umask, cwd or root directory shared with other threads. The thread will no longer be indistinguishable from other threads which run goroutines. Returning from the goroutine while locked to the thread signals to the Go runtime to terminate the thread ([or wedge, if it's the main thread](https://github.com/golang/go/blob/895664482c0ebe5cec4a6935615a1e9610bbf1e3/src/runtime/proc.go#L1476-L1483)) for situations exactly like this one. ([This article has some good information as well.](https://www.weave.works/blog/linux-namespaces-golang-followup)) Just be aware that [terminating threads can have some undesirable interactions with `(syscall.SysProcAttr).Pdeathsig`](https://go.dev/issue/27505) which need to be accounted for. It's straightforward enough to deal with once you know how what's happening and why, e.g.: https://github.com/moby/moby/pull/44215.

Reply:
thanks for the information! updated as suggested